### ISSUE TITLE
BatchNorm: C++ Frontend ignoring running_mean and running_var

### ISSUE THREAD

**Issue by McChicken93** [pytorch/pytorch#21679]
## 🐛 Bug

When using the PyTorch C++ frontend, the BatchNorm-Modules ignore running_mean and running_var values.

## To Reproduce

```c++
#include <iostream>
#include <torch/torch.h>


int main(int argc, char** argv)
{
  float data[3][8] = {{1,1,3,3,2,2,4,4},
                      {1,1,1,1,1,1,1,1},
                      {1,1,1,1,1,1,1,1}};
  auto m = torch::nn::BatchNorm(torch::nn::BatchNormOptions(2));
  m->weight = torch::from_blob(&data[0][0],{2});
  m->bias = torch::from_blob(&data[0][2],{2});
  m->running_mean = torch::from_blob(&data[0][4],{2});
  m->running_var = torch::from_blob(&data[0][6],{2});

  auto test_input = torch::from_blob(data[1],{1,2,2,2});

  std::cout << "Input:\n" << test_input << "\n\n\n\n";
  std::cout << "Weight:\n" << m->weight << "\n\n\n\n";
  std::cout << "Bias:\n" << m->bias << "\n\n\n\n";
  std::cout << "Running Mean:\n" << m->running_mean << "\n\n\n\n";
  std::cout << "Running Var:\n" << m->running_var << "\n\n\n\n";
  std::cout << "Feature Mean:\n" << test_input.slice(0,0,1).mean() << "\n\n\n\n";
  std::cout << "Feature Var:\n" << test_input.slice(0,0,1).std()<< "\n\n\n\n";

  std::cout << "Forward:\n" << m->forward(test_input) << "\n\n\n\n";
  std::cout << "Forward:\n" << m->pure_forward(test_input, m->running_mean, m->running_var) << "\n\n\n\n";

}
```

Returns for both forwards a value of [3, 3]

## Expected behavior

According to the Pytorch documentation https://pytorch.org/docs/stable/nn.html#torch.nn.BatchNorm2d the value should be [2.5, 2.5].

## Environment

 - PyTorch Version (e.g., 1.0): 1.1.0
 - OS (e.g., Linux): Ubuntu 16.04
 - How you installed PyTorch (`conda`, `pip`, source): source
 - Build command you used (if compiling from source): CMAKE_INSTALL_PREFIX='/usr/local/' CFLAGS=' -D_GLIBCXX_USE_CXX11_ABI=1 ' USE_OPENCV=1 USE_CUDA=1 MAX_JOBS=5 python setup.py install
 - Python version: 3.5.2
 - CUDA/cuDNN version: CUDA 10.0,  cuDNN 7.5
 - GPU models and configuration: 
 - Any other relevant information:

**Comment by pietern** [comment]
I don't know why this is happening myself, but marked as "high priority" so someone else will take a look.

**Comment by ezyang** [comment]
The example looks like it is giving the correct result to me.

```
import torch
import torch.nn as nn

m = nn.BatchNorm2d(2)
m.weight.data.fill_(1)
m.bias.data.fill_(3)
m.running_mean.data.fill_(2)
m.running_var.data.fill_(4)

test_input = torch.ones(1, 2, 2, 2)

print(m(test_input))
print(m(test_input))
```

gives

```
tensor([[[[3., 3.],
          [3., 3.]],

         [[3., 3.],
          [3., 3.]]]], grad_fn=<NativeBatchNormBackward>)
tensor([[[[3., 3.],
          [3., 3.]],

         [[3., 3.],
          [3., 3.]]]], grad_fn=<NativeBatchNormBackward>)
```

Your problem is probably that you forgot to set the module to `eval()` mode.

```
(/scratch/ezyang/pytorch-tmp-env) ezyang@devfair040:/scratch/ezyang/pytorch-tmp$ cat foo.py
import torch
import torch.nn as nn

m = nn.BatchNorm2d(2).eval()
m.weight.data.fill_(1)
m.bias.data.fill_(3)
m.running_mean.data.fill_(2)
m.running_var.data.fill_(4)

test_input = torch.ones(1, 2, 2, 2)

print(m(test_input))
print(m(test_input))
(/scratch/ezyang/pytorch-tmp-env) ezyang@devfair040:/scratch/ezyang/pytorch-tmp$ python foo.py 
tensor([[[[2.5000, 2.5000],
          [2.5000, 2.5000]],

         [[2.5000, 2.5000],
          [2.5000, 2.5000]]]], grad_fn=<NativeBatchNormBackward>)
tensor([[[[2.5000, 2.5000],
          [2.5000, 2.5000]],

         [[2.5000, 2.5000],
          [2.5000, 2.5000]]]], grad_fn=<NativeBatchNormBackward>)
```

You can do this by calling `eval()` on your C++ module.

**Comment by ezyang** [comment]
By the way, replacing parameters this way:

```

  m->weight = torch::from_blob(&data[0][0],{2});
  m->bias = torch::from_blob(&data[0][2],{2});
  m->running_mean = torch::from_blob(&data[0][4],{2});
  m->running_var = torch::from_blob(&data[0][6],{2});
```

will break `parameters()` machinery. In general, don't assign over members in modules in the C++ API.
